### ISSUE TITLE
Local Development

### ISSUE THREAD

**Issue by ethernetdan** [redspread/spread#95]
A major barrier to the adoption of Kubernetes is the amount work that is required to setup an environment for rapid development.

We would like to introduce a feature of spread which would manage a Kube cluster on behalf of a user. It would have tight integration with several development specific commands (ie. `spread build`). A kubectl context will be created for the cluster for ease of use.

It would be implemented and function very similarly to [monokube](https://github.com/polvi/monokube) with tight integration with spread.

**Comment by ethernetdan** [comment]
#### Commands
**spread start**
* Run all Kubernetes components + etcd in goroutines
* create appropriate kubectl context
* Setup etcd storage

**spread stop**
* Kills all running containers
* Stops Kubernetes + etcd

**spread reset**
* Clears etcd
* Should confirm before action

**Comment by ethernetdan** [comment]
### Components
#### etcd
* Persist data between usages
    * Use folder in home directory for storage (ie. `~/.localkube`)
* Cleared using reset command

#### apiserver
* Allow unauthenticated & insecure requests (for easy testing)
    * this might change in the future
* Should be accessible locally on the computer
    * preferably using a human readable hostname (ie. `localkube`)
* Connects to local etcd instance
* Initially no support for cloud provider functionalities (load balancing, volumes)

#### controller-manager
* Default config
* Setup to use local apiserver (above)

#### scheduler
* Default config
* Setup to use local apiserver (above)

#### DNS
* Create Kubernetes objects at cluster startup:
    * [ReplicationController](https://github.com/kubernetes/kubernetes/blob/d644062/cluster/addons/dns/skydns-rc.yaml.in) running [kube2sky](https://github.com/kubernetes/kubernetes/tree/master/cluster/addons/dns/kube2sky) and [SkyDNS](https://github.com/skynetservices/skydns)
    * [Service](https://github.com/kubernetes/kubernetes/blob/d644062/cluster/addons/dns/skydns-svc.yaml.in) for DNS Queries with static IP
* Ideally should be usable from host operating system

#### Networking
* We want to use some form of network overlay in order to enable 1 Pod-1 IP.
    * Unfortunately, this probably requires performing actions before the Docker daemon starts
    * This is would mean we couldn't rely on any working Docker configuration

#### kubelet
* Default config
* Setup to use local apiserver (above)

**Comment by ethernetdan** [comment]
I've done some thinking and research on this the past few days and have these thoughts:

* Adding the entire control plane plus the kubelet and etcd as dependencies will drastically increase spread's already large executable size. For comparison, monokube weighs in at a hefty 80+ MB and would have less functionality. Let's create a new executable specifically for managing a local development cluster which would contain the dependencies instead. I'm unsure whether or not to run the components inside a container or not.

* It looks like running a solo instance of [weave](https://github.com/weaveworks/weave) will be able to satisfy our networking requirements without requiring a modified Docker environment or daemon restarts. It operates by proxying requests to the Docker daemon allowing it to correctly configure any containers created. All components should interface with Docker through this proxy. It looks like it shouldn't be a problem to run this in a container itself.

**Comment by ethernetdan** [comment]
I'm thinking we go with having the cluster running from a single executable so that it can be easily killed (single process) and it's components are easier to configure. This executable will run out of a container which is able to be started with spread to make Kubernetes versioning easier.

- [ ] LocalKube Image
	- [ ] Weave Router
	- [ ] Weave Proxy (for Docker daemon)
	- [ ] OmniKube executable
		- [ ] etcd
		- [ ] apiserver
		- [ ] controller-manager
		- [ ] scheduler
		- [ ] kubelet (configured for weave proxy)
- [ ] DNS (in its own container or potentially part of omnikube)

**Comment by ethernetdan** [comment]
Mistake.

**Comment by jbowen93** [comment]
I'm curious what the motivation is behind running K8s locally as opposed to a tightly integrated VM workflow. My initial thoughts were since most developers seem to be on OSX and all K8s clusters will be run in prod on a linux box it would be best to just have a streamlined docker/kubernetes workflow with a local vm.

**Comment by thecolorblue** [comment]
@ethernetdan, looking good. +1 for running in a vm and +1 for splitting the whole idea into multiple containers.

In the localKube/DNS setup you described, are you describing two different localKube containers and a DNS container?

**Comment by ethernetdan** [comment]
@jbowen93: mostly to provide support for running clusters on Linux and remote Docker daemons. I'm thinking we provide that functionality on non-Linux platforms by integrating with [docker-machine](https://github.com/docker/machine/tree/b53f020/libmachine). The goal would be to be able to `spread cluster start` and having a working cluster regardless of what platform is being used. Thoughts?

@thecolorblue: What do you want out of having the control plane running across multiple containers? I am a bit wary of having to work around the boundaries of containers and cluttering the Docker daemon.

Oops, that was a typo (fixed above). There's only supposed to one localkube container.

**Comment by thecolorblue** [comment]
@ethernetdan, looks better :). I knew you did not mean to repeat yourself but I couldn't tell what you did mean. 

I'm not sure what you mean by control plane. 

I'm only just getting started with kubernetes, but I do know that setting up multiple docker containers that talk to each other seperates the concerns of my application perfectly. What is different about kubernetes so that pattern would not work?

**Comment by jsquirrelz** [comment]
@thecolorblue here's a brief overview of the control plane [concept]( https://github.com/kubernetes/kubernetes/blob/master/docs/design/architecture.md#the-kubernetes-control-plane). 

Also +1 to using [docker-machine](https://github.com/docker/machine). Really don't see a reason to use any other VM setup.

Definitely want to make sure I understand this correctly: 

- the LocalKube image is going to be the master node/control plane for the local kubernetes cluster (via OmniKube?).
- Weave and the DNS support to primarily allow spread to manage local deployment?

But great work, I definitely think getting local development/deployment working via `spread cluster start` will dramatically reduce the barriers to adoption.

**Comment by ethernetdan** [comment]
@thecolorblue: The environment would definitely allow for users to run multiple containers within a cluster, I'm thinking just the backend components that run Kubernetes would be run out of a single container. Does that resolve your concern?

@jsquirrelz: Definitely could be more clear, sorry for the confusion:
* Yes, the LocalKube image would contain everything required for a working cluster (control plane + kubelet). "OmniKube" was my [poor choice of name](https://www.simplivity.com/omnicube-product-family/) for the executable which contains these components. We should choose a new name, any suggestions?
* I was thinking Weave would be used to provide a network overlay for the 1 IP <-> 1 Pod mapping that is [expected by Kubernetes](https://github.com/kubernetes/kubernetes/blob/dae84f5/docs/design/networking.md#pod-to-pod). The DNS support mentioned above is for "in-cluster" [resolution of Services](https://github.com/kubernetes/kubernetes/tree/a3c00aa/cluster/addons/dns).

Does that make sense?

In addition to providing these services within the cluster, it would be nice if the cluster's network and DNS were accessible from the host machine for easy debugging. I'm not sure what the best solution is there.

**Comment by jsquirrelz** [comment]
Gotcha. I'd say keep it simple like LocalKube or DevKube (because it's just for local development, right?). And interesting; I never knew that DNS was an addon for Kubernetes.  Always thought that was just 'out-of-the-box' functionality (then again, only exposure is through GKE). But yeah that makes a lot more sense to me, thanks.

And couldn't you setup Weave to include the host machine in the topology? Or am I missing the point there? You could potentially setup a dedicated pod to run in debug mode that would expose some debugging tools as services. Not sure if that's necessary though. Probably an easier way.

**Comment by rata** [comment]
Hi @ethernetdan. I like the proposal, great that weave solves the networking easily. What I'm not sure what the plan is how to handle the code in localhost. I.e. would it be mounted through a docker volume? (then, you just edit as always, and a container sees it too)

Or is there some other plan?

**Comment by ethernetdan** [comment]
@jsquirrelz: Unfortunately weave is linux only, which seems to be the case for [all the overlays I know of that work with Kube](https://github.com/kubernetes/kubernetes/blob/dae84f5/docs/design/networking.md#other-implementations).

Interesting idea about having a pod for debug tools, I like the idea of shipping those tool independently so users aren't required to install what they don't need. We could ship different pods which offered different cluster services that might be useful in dev. For example, OpenVPN to solve the host to cluster connectivity issue.

@rata: That strategy would definitely work, it could be accomplished by using a HostPathVolumeSource set to a path in the shared [host directory](https://docs.docker.com/v1.8/userguide/dockervolumes/#mount-a-host-directory-as-a-data-volume) setup by docker-machine on OS X and Windows or the actual path on Linux. This is such a commonly used functionality in development though it would be nice if that was easier to use (support relative host path, not having to put both a VolumeSource and a VolumeMount).

In the workflow I'm recommending users, build using the same docker-machine the cluster is running on so that all images can be immediately used in Kube without pushing.

**Comment by rata** [comment]
@ethernetdan but what is the strategy that you propose? I mean, I was thinking out loud. I want to know which will be the easy way to develop on localhost using bedspread :)

**Comment by jbowen93** [comment]
Some links to current projects working on kubernetes local development. Note both of these are made by me but forked from other projects. 

[Docker-compose-kubernetes](https://github.com/30x/docker-compose-kubernetes)

[Vagrant Single-Node](https://github.com/30x/Dev_Setup) 

The docker-compose-kubernetes setup is similar to what I believe is being discussed here. Where the user simply runs docker-machine. However in this case the kubernetes environment is created by using a docker-compose file to spin up the relevant containers.

The Vagrant setup works by spinning up a single-node CoreOS VM running docker locally with all of the necessary kubernetes components. For ease of use the docker commands are exposed on the vm over tcp and can be reached by modifying the clients `DOCKER_HOST` environment variable.

**Comment by rata** [comment]
Ping? :)

**Comment by ethernetdan** [comment]
This conversation is getting a little cluttered so I've broken it down into several issues in a new [localkube](https://github.com/redspread/localkube) repo. There is an issue for each Kubernetes component required summarizing some of the discussion here. Also I got started with the project and implemented etcd!

I'll keep this open until things settle a bit with the goal of moving all discussion into localkube issues.

@rata: I wrote a little about my thoughts around developer workflow with localkube in redspread/localkube#9. Any feedback would be greatly appreciated.
